enhancements/microshift/y-2-upgrades.md

```diff
@@ -0,0 +1,223 @@
+
+If, in the future, MicroShift does need to support multi-node
+deployments there will be many other aspects of deployment and upgrade
+to consider, in addition to the version skew problem. We can envision
+implementing a process similar to what OpenShift uses, where the
+control plane and workers are updated using separate steps.
+
+If an upgrade fails, even after a complex data migration, MicroShift's
+rollback process is to discard the new database and restore the old
+version from a backup before continuing. This ensures that an old
+version of the software matches the older database (file format,
+schema, and content).
+
+### Drawbacks
+
+The main drawback to implementing this enhancement is the increased
+test matrix for upgrades. We can automate those tests to minimize the
+impact.
+
+## Design Details
+
+### Test Plan
+
+We will add an automated test to CI to deploy 4.Y-2 and update to 4.Y
+using the latest published packages of 4.Y-2 and testing the "source"
+version (HEAD of the branch or the pull request content) of 4.Y. This
+ensures that every package we build can be continuously upgraded to
+the latest version of the source.
+
+The QE team will need to perform similar tests using the 4.Y-2 and 4.Y
+packages built by the release team.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+The mechanics of upgrade and rollback for MicroShift do not change as
+part of this work.
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+N/A
+
+#### Support Procedures
+
+N/A
+
+## Implementation History
+
+* https://github.com/openshift/microshift/pull/2952
+
+## Alternatives
+
+We could limit the ability to skip versions so that it is possible to
+go from an even version (EUS) to the next odd or even version, but not
+allow moving from an odd (non-EUS) version to the next odd version
+(4.14 to 4.16 would be OK, but 4.15 to 4.17 would not). This would
+make the version checking logic more complicated and would introduce
+opportunities for that skip-level upgrade process to be broken in a
+non-EUS version so that it has to be fixed before the next EUS
+release. By allowing skipping 1 of any type of version, we test the
+feature continuously and avoid those issues.
```

Comment:
Agreed. And it would simplify docuementation, create a more "linear" usage expierience. So from that perspective, I like the approach. Still, I wonder if test complexity would explode and be too much for the team to handle.

Reply:
I don't think this is going to be a problem. We add 1 new test scenario to the existing suite. In fact, as part of the actually implementation I _removed_ 2 other scenarios that were no longer relevant, so we have a net negative effect on the number of tests.